File: src/api/dcps/sacs/code/DDS/OpenSplice/Topic.cs
/*
 *                         OpenSplice DDS
 *
 *   This software and documentation are Copyright 2006 to TO_YEAR PrismTech
 *   Limited, its affiliated companies and licensors. All rights reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

using System;
using System.Runtime.InteropServices;
using System.Collections.Generic;
using System.Diagnostics;
using DDS;
using DDS.OpenSplice.OS;
using DDS.OpenSplice.Database;
using DDS.OpenSplice.CustomMarshalers;
using DDS.OpenSplice.Kernel;
using DDS.OpenSplice.kernelModuleI;

namespace DDS.OpenSplice
{
    internal class Topic : Entity, ITopic, ITopicDescriptionImpl
    {
//        private readonly TopicListenerHelper listenerHelper;
        private DomainParticipant domainParticipant;
        private TypeSupport typeSupport;
        private string topicName;
        private string typeName;
        private string topicExpression;
        private uint nrUsers = 0;

        internal Topic()
        {
        }

        internal ReturnCode init (
                IntPtr uTopic,
                DomainParticipant participant,
                string topicName,
                string typeName,
                TypeSupport ts)
        {
            MyDomainId = participant.MyDomainId;

            ReturnCode result = base.init (uTopic);

            if (result == DDS.ReturnCode.Ok) {
                this.domainParticipant = participant;
                this.topicName = topicName;
                this.typeName = typeName;
                this.typeSupport = ts;
                this.topicExpression = "select * from " + topicName;
            } else {
                ReportStack.Report (result, "Could not create Topic" + topicName + ".");
            }


            return result;
        }

        internal override ReturnCode wlReq_deinit()
        {
            ReturnCode result = DDS.ReturnCode.Ok;
            if (nrUsers > 0)
            {
                result = DDS.ReturnCode.PreconditionNotMet;
                ReportStack.Report(result, "Topic \"" + topicName + "\" still in use by " + nrUsers + " Writers/Readers.");
            }
            else
            {
                ITopicListener tpListener = listener as ITopicListener;
                if (tpListener != null)
                {
                    this.SetListener(tpListener,(DDS.StatusKind)0);
                }
                this.DisableCallbacks();
                result = base.wlReq_deinit();
                if (result == DDS.ReturnCode.Ok)
                {
                    this.domainParticipant = null;
                    this.topicName = null;
	                this.typeName = null;
	                this.typeSupport = null;
	                this.topicExpression = null;
                }
            }
            return result;
        }

        string ITopicDescriptionImpl.rlReq_TopicExpression
        {
            get
            {
                return topicExpression;
            }
        }

        string[] ITopicDescriptionImpl.rlReq_TopicExpressionParameters
        {
            get
            {
                return null;
            }
        }

        void ITopicDescriptionImpl.wlReq_IncrNrUsers()
        {
            nrUsers++;
        }

        ReturnCode ITopicDescriptionImpl.DecrNrUsers()
        {
            ReturnCode result = DDS.ReturnCode.AlreadyDeleted;

            lock(this)
            {
                if (this.rlReq_isAlive)
                {
                    nrUsers--;
                    result = DDS.ReturnCode.Ok;
                }
            }
            return result;
        }

        TypeSupport ITopicDescriptionImpl.rlReq_TypeSupport
        {
            get
            {
                return typeSupport;
            }
            set
            {
                typeSupport = value;
            }
        }

        internal override void NotifyListener(Entity source, V_EVENT triggerMask, DDS.OpenSplice.Common.EntityStatus status)
        {
            ITopicListener tpListener = listener as ITopicListener;
            if (tpListener != null)
            {
                DDS.OpenSplice.Common.TopicStatus topicStatus = status as DDS.OpenSplice.Common.TopicStatus;

	            if ((triggerMask & V_EVENT.INCONSISTENT_TOPIC) == V_EVENT.INCONSISTENT_TOPIC)
	            {
	                tpListener.OnInconsistentTopic(source as ITopic, topicStatus.InconsistentTopic);
	            }

	            /*if ((triggerMask & V_EVENT.ALL_DATA_DISPOSED) == V_EVENT.ALL_DATA_DISPOSED)
	            {
	                IDomainParticipantListener dpExtListener = listener as IExtDomainParticipantListener;
	                Debug.Assert(dpExtListener != null);
	                dpExtListener.OnAllDataDisposed(source as ITopic);
	            }*/
	        }
        }

        internal ReturnCode ValidateFilter(string filterExpression, string[] filterParameters)
        {
            ReturnCode result = DDS.ReturnCode.BadParameter;
            int length = (filterParameters != null) ? filterParameters.Length : 0;
            IntPtr qExpr = IntPtr.Zero;
            IntPtr qParams = IntPtr.Zero;

            ReportStack.Start();
            if (length < 100)
            {
                qExpr = DDS.OpenSplice.Kernel.Parser.parse(filterExpression);
                if (qExpr != IntPtr.Zero && length > 0)
                {
                    SequenceStringToCValueArrMarshaler.CopyIn(filterParameters, ref qParams);
                }
                else
                {
                    ReportStack.Report(result, "ContentFilteredTopic cannot be created. Invalid expression: " + filterExpression);
                }
            }
            else
            {
                ReportStack.Report(result, "ContentFilteredTopic cannot be created. Too many parameters " + length + " (max = 99)");
            }

            if (qExpr != IntPtr.Zero)
            {
                if (User.Topic.ContentFilterValidate2(rlReq_UserPeer, qExpr, qParams) != 0)
                {
                    result = DDS.ReturnCode.Ok;
                }
                else
                {
                    ReportStack.Report(result, "ContentFilteredTopic cannot be created. Filter invalid: " + filterExpression);
                }
                if (qParams != IntPtr.Zero)
                {
                    SequenceStringToCValueArrMarshaler.CleanupIn(ref qParams, length);
                }
                q.dispose(qExpr);
            }
            ReportStack.Flush(this, result != ReturnCode.Ok);
            return result;
        }

        public ITopicListener Listener
        {
            get
            {
                bool isAlive;
                ITopicListener tpListener = null;

                ReportStack.Start();
                lock(this)
                {
                    isAlive = this.rlReq_isAlive;
                    if (isAlive)
                    {
                        tpListener = rlReq_Listener as ITopicListener;
                    }
                }
                ReportStack.Flush(this, !isAlive);

                return tpListener;
            }
        }

        public ReturnCode SetListener(ITopicListener listener, StatusKind mask)
        {
            ReturnCode result = DDS.ReturnCode.AlreadyDeleted;

            ReportStack.Start();
            lock(this)
            {
                if (this.rlReq_isAlive)
                {
                    result = wlReq_SetListener(listener, mask);
                }
            }
            ReportStack.Flush(this, result != ReturnCode.Ok);

            return result;
        }

        public ReturnCode GetInconsistentTopicStatus(ref InconsistentTopicStatus status)
        {
            ReturnCode result = DDS.ReturnCode.AlreadyDeleted;
            ReportStack.Start();
            if (this.rlReq_isAlive)
            {
                if (status == null) status = new InconsistentTopicStatus();
                GCHandle statusGCHandle = GCHandle.Alloc(status, GCHandleType.Normal);
                result = uResultToReturnCode(
                        User.Topic.GetInconsistentTopicStatus(
                                rlReq_UserPeer, 1, InconsistentTopicStatusMarshaler.CopyOut, GCHandle.ToIntPtr(statusGCHandle)));
                status = statusGCHandle.Target as InconsistentTopicStatus;
                statusGCHandle.Free();
            }
            ReportStack.Flush(this, result != ReturnCode.Ok);
            return result;
        }

        public ReturnCode GetQos(ref TopicQos qos)
        {
            IntPtr userQos = IntPtr.Zero;
            ReturnCode result = DDS.ReturnCode.AlreadyDeleted;
            ReportStack.Start();
            if (this.rlReq_isAlive)
            {
                result = uResultToReturnCode(
                        User.Topic.GetQos(rlReq_UserPeer, ref userQos));
                if (result == ReturnCode.Ok)
                {
                    using (OpenSplice.CustomMarshalers.TopicQosMarshaler marshaler =
                            new OpenSplice.CustomMarshalers.TopicQosMarshaler(userQos, true))
                    {
                        marshaler.CopyOut(ref qos);
                    }
                }
                else
                {
                    ReportStack.Report(result, "Could not copy TopicQos.");
                }
            }
            ReportStack.Flush(this, result != ReturnCode.Ok);

            return result;
        }

        public ReturnCode SetQos(TopicQos qos)
        {
            ReturnCode result = DDS.ReturnCode.AlreadyDeleted;
            ReportStack.Start();
            if (this.rlReq_isAlive)
            {
                result = QosManager.checkQos(qos);
                if (result == DDS.ReturnCode.Ok)
                {
                    using (OpenSplice.CustomMarshalers.TopicQosMarshaler marshaler =
                        new OpenSplice.CustomMarshalers.TopicQosMarshaler())
                    {
                        result = marshaler.CopyIn(qos);
                        if (result == ReturnCode.Ok)
                        {
                            result = uResultToReturnCode(
                                    User.Topic.SetQos(rlReq_UserPeer, marshaler.UserPtr));
                            if (result != ReturnCode.Ok)
                            {
                                ReportStack.Report(result, "Could not apply TopicQos.");
                            }
                        }
                    }
                }
            }
            ReportStack.Flush(this, result != ReturnCode.Ok);

            return result;
        }

        /// <summary>
        /// This property returns the registered name of the data type associated with the TopicDescription
        /// (inherited from TopicDescription)
        /// </summary>
        public string TypeName
        {
            get
            {
                bool isAlive;
                string name = null;

                ReportStack.Start();
                lock(this)
                {
                    isAlive = this.rlReq_isAlive;
                    if (isAlive)
                    {
                        name = typeName;
                    }
                }
                ReportStack.Flush(this, !isAlive);

                return name;
            }
        }

        /// <summary>
        /// This property returns the name used to create the TopicDescription.
        /// (inherited from TopicDescription)
        /// </summary>
        public string Name
        {
            get
            {
                bool isAlive;
                string name = null;

                ReportStack.Start();
                lock(this)
                {
                    isAlive = this.rlReq_isAlive;
                    if (isAlive)
                    {
                        name = topicName;
                    }
                }
                ReportStack.Flush(this, !isAlive);

                return name;
            }
        }

        /// <summary>
        /// This property returns the DomainParticipant associated with the TopicDescription or
        /// a null DomainParticipant.
        /// (inherited from TopicDescription)
        /// </summary>
        public IDomainParticipant Participant
        {
            get
            {
                bool isAlive;
                DomainParticipant participant = null;

                ReportStack.Start();
                lock(this)
                {
                    isAlive = this.rlReq_isAlive;
                    if (isAlive)
                    {
                        participant = domainParticipant;
                    }
                }
                ReportStack.Flush(this, !isAlive);

                return participant;
            }
        }
    }
}

File: src/api/dcps/c++/common/code/Topic.cpp
/*
 *                         Vortex OpenSplice
 *
 *   This software and documentation are Copyright 2006 to TO_YEAR ADLINK
 *   Technology Limited, its affiliated companies and licensors. All rights
 *   reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */
#include "v_kernelParser.h"
#include "u_user.h"
#include "u_observable.h"
#include "DomainParticipant.h"
#include "TopicDescription.h"
#include "Topic.h"
#include "StatusUtils.h"
#include "Constants.h"
#include "QosUtils.h"
#include "ReportUtils.h"

using namespace DDS::OpenSplice::Utils;


DDS::Boolean
DDS::OpenSplice::set_topic_listener_mask (
  DDS::Object_ptr element,
  void *arg)
{
    DDS::OpenSplice::Topic *topic = dynamic_cast<DDS::OpenSplice::Topic *>(element);
    DDS::StatusMask *mask = (DDS::StatusMask *)arg;
    DDS::Boolean result = false;

    if (topic) {
        result = topic->set_participant_listener_mask(*mask) == DDS::RETCODE_OK;
    }
    return result;
}

DDS::OpenSplice::Topic::Topic() :
    DDS::OpenSplice::Entity(DDS::OpenSplice::TOPIC),
    topicListenerInterest(0),
    participantListenerInterest(0)
{
}

DDS::OpenSplice::Topic::~Topic()
{
}

DDS::ReturnCode_t
DDS::OpenSplice::Topic::init(
    const u_topic uTopic,
    DDS::OpenSplice::DomainParticipant *participant,
    const DDS::Char *topic_name,
    const DDS::Char *type_name,
    DDS::OpenSplice::TypeSupportMetaHolder *meta_holder)
{
    DDS::ReturnCode_t result;
    result = this->nlReq_init(uTopic, participant, topic_name, type_name, meta_holder);
    return result;
}

DDS::ReturnCode_t
DDS::OpenSplice::Topic::nlReq_init(
    const u_topic uTopic,
    DDS::OpenSplice::DomainParticipant *participant,
    const DDS::Char *topic_name,
    const DDS::Char *type_name,
    DDS::OpenSplice::TypeSupportMetaHolder *meta_holder)
{
    DDS::ReturnCode_t result;
    DDS::Char *expression;
    DDS::ULong length;
    const DDS::Char *prefix = "select * from ";

    assert (participant != NULL);
    assert (topic_name != NULL);
    assert (type_name != NULL);
    assert (uTopic != NULL);

    result = DDS::OpenSplice::Entity::nlReq_init(u_entity(uTopic));

    if (result == DDS::RETCODE_OK) {
        length = strlen(prefix) + strlen(topic_name) + 1;
        expression = DDS::string_alloc(length);
        if ( expression ) {
            snprintf(expression, length, "%s%s", prefix, topic_name);
            result = DDS::OpenSplice::TopicDescription::nlReq_init(participant,
                                                              topic_name,
                                                              type_name,
                                                              expression,
                                                              meta_holder);
            DDS::string_free(expression);
            setDomainId(participant->getDomainId());
        } else {
            result = DDS::RETCODE_OUT_OF_RESOURCES;
            CPP_REPORT(result, "Could not create Topic '%s'.", topic_name);
        }
    }

    return result;
}

DDS::ReturnCode_t
DDS::OpenSplice::Topic::wlReq_deinit()
{
    DDS::ReturnCode_t result;

    result = DDS::OpenSplice::TopicDescription::wlReq_deinit();
    if (result == DDS::RETCODE_OK) {
        this->disable_callbacks();

        result = DDS::OpenSplice::Entity::wlReq_deinit();
    }
    return result;
}

static v_result
copy_inconsistent_topic_status (
    c_voidp info,
    c_voidp arg)
{
    struct v_inconsistentTopicInfo *from;
    ::DDS::InconsistentTopicStatus *to;

    from = (struct v_inconsistentTopicInfo *)info;
    to = (::DDS::InconsistentTopicStatus *)arg;

    to->total_count = from->totalCount;
    to->total_count_change = from->totalChanged;

    return V_RESULT_OK;
}

DDS::ReturnCode_t
DDS::OpenSplice::Topic::get_inconsistent_topic_status (
    DDS::InconsistentTopicStatus &a_status
) THROW_ORB_EXCEPTIONS
{
    DDS::ReturnCode_t result;
    u_result uResult;

    CPP_REPORT_STACK();

    result = this->check ();
    if (result == DDS::RETCODE_OK) {
        uResult = u_topicGetInconsistentTopicStatus(
                      u_topic(DDS::OpenSplice::Entity::rlReq_get_user_entity()),
                      TRUE,
                      copy_inconsistent_topic_status,
                      &a_status);
        result = this->uResultToReturnCode(uResult);
        if (result != DDS::RETCODE_OK) {
            CPP_REPORT(result, "Could not get topic status.");
        }
    }

    CPP_REPORT_FLUSH(this, result != DDS::RETCODE_OK);

    return result;
}

DDS::ReturnCode_t
DDS::OpenSplice::Topic::validate_filter (
    const DDS::Char *filter_expression,
    const DDS::StringSeq &filter_parameters
) THROW_ORB_EXCEPTIONS
{
    DDS::ReturnCode_t result;
    q_expr expr = NULL;
    c_value *params = NULL;
    u_topic uTopic = NULL;
    int n, length;

    CPP_REPORT_STACK();

    result = DDS::RETCODE_BAD_PARAMETER;
    length = filter_parameters.length();
    if ((length >= 0) && (length < 100)) {
        expr = v_parser_parse(filter_expression);
        if (expr != NULL) {
            if(length > 0) {
                params = (c_value *)os_malloc(length * sizeof(struct c_value));
                for (n=0;n<length;n++) {
                    const c_string param = (const c_string) (const char *)(filter_parameters[n]);
                    params[n] = c_stringValue(param);
                }
            }
        } else {
            CPP_REPORT(result, "filter_expression '%s' is invalid", filter_expression);
        }
    } else {
        CPP_REPORT(result, "Invalid number of filter_parameters '%d', maximum is 99", length);
    }

    if (expr) {
        uTopic = u_topic(DDS::OpenSplice::Entity::rlReq_get_user_entity());
        if (u_topicContentFilterValidate2(uTopic, expr, params, length)) {
            result = DDS::RETCODE_OK;
        } else {
            CPP_REPORT(result, "filter_expression '%s' is invalid.", filter_expression);
        }
        q_dispose(expr);
        os_free(params);
    }

    CPP_REPORT_FLUSH(this, result != DDS::RETCODE_OK);

    return result;
}

static v_result
copy_all_data_disposed_status (
    c_voidp info,
    c_voidp arg)
{
    struct v_allDataDisposedInfo *from;
    ::DDS::AllDataDisposedTopicStatus *to;

    from = (struct v_allDataDisposedInfo *)info;
    to = (::DDS::AllDataDisposedTopicStatus *)arg;

    to->total_count = from->totalCount;
    to->total_count_change = from->totalChanged;

    return V_RESULT_OK;
}

DDS::ReturnCode_t
DDS::OpenSplice::Topic::get_all_data_disposed_topic_status (
    DDS::AllDataDisposedTopicStatus & a_status
) THROW_ORB_EXCEPTIONS
{
    DDS::ReturnCode_t result;
    u_result uResult;

    CPP_REPORT_STACK();

    result = this->check ();
    if (result == DDS::RETCODE_OK) {
        uResult = u_topicGetAllDataDisposedStatus(
                      u_topic(DDS::OpenSplice::Entity::rlReq_get_user_entity()),
                      TRUE,
                      copy_all_data_disposed_status,
                      &a_status);
        result = this->uResultToReturnCode(uResult);
        if (result != DDS::RETCODE_OK) {
            CPP_REPORT(result, "Could not get topic status.");
        }
    }

    CPP_REPORT_FLUSH(this, result != DDS::RETCODE_OK);

    return result;
}

DDS::ReturnCode_t
DDS::OpenSplice::Topic::get_qos (
    DDS::TopicQos & qos
) THROW_ORB_EXCEPTIONS
{
    DDS::ReturnCode_t result;
    u_topic uTopic;
    u_topicQos uTopicQos;
    u_result uResult;

    CPP_REPORT_STACK();

    if (&qos == &TOPIC_QOS_DEFAULT) {
        result = DDS::RETCODE_BAD_PARAMETER;
        CPP_REPORT(result, "QoS 'TOPIC_QOS_DEFAULT' is read-only.");
    } else {
        result = this->check ();
    }

    if (result == DDS::RETCODE_OK) {
        uTopic = u_topic (rlReq_get_user_entity ());
        uResult = u_topicGetQos (uTopic, &uTopicQos);
        result = uResultToReturnCode (uResult);
        if (result == DDS::RETCODE_OK) {
            result = DDS::OpenSplice::Utils::copyQosOut (uTopicQos, qos);
            u_topicQosFree(uTopicQos);
        } else {
            CPP_REPORT(result, "Could not copy TopicQos.");
        }
    }

    CPP_REPORT_FLUSH(this, result != DDS::RETCODE_OK);

    return result;
}

DDS::ReturnCode_t
DDS::OpenSplice::Topic::set_qos (
    const DDS::TopicQos & qos
) THROW_ORB_EXCEPTIONS
{
    DDS::ReturnCode_t result;
    DDS::TopicQos topicQos;
    DDS::TopicQos *topicQos_ptr;
    u_topic uTopic;
    u_topicQos uTopicQos;
    u_result uResult;

    CPP_REPORT_STACK();

    if (&qos == &TOPIC_QOS_DEFAULT) {
        topicQos_ptr = NULL;
        result = DDS::RETCODE_OK;
    } else {
        topicQos_ptr = const_cast<DDS::TopicQos *>(&qos);
        result = DDS::OpenSplice::Utils::qosIsConsistent (qos);
    }

    if (result == DDS::RETCODE_OK) {
        uTopicQos = u_topicQosNew (NULL);
        if (uTopicQos == NULL) {
            result = DDS::RETCODE_OUT_OF_RESOURCES;
            CPP_REPORT(result, "Could not copy TopicQos.");
        }
    }

    if (result == DDS::RETCODE_OK) {
        result = this->write_lock ();
        if (result == DDS::RETCODE_OK) {
            if (topicQos_ptr == NULL) {
                result = DDS::OpenSplice::TopicDescription::participant->get_default_topic_qos (topicQos);
                topicQos_ptr = &topicQos;
            }

            if (result == DDS::RETCODE_OK) {
                result = DDS::OpenSplice::Utils::copyQosIn (
                                                    *topicQos_ptr, uTopicQos);
                if (result == DDS::RETCODE_OK) {
                    uTopic = u_topic (rlReq_get_user_entity ());
                    uResult = u_topicSetQos (uTopic, uTopicQos);
                    result = uResultToReturnCode (uResult);
                    if (result != DDS::RETCODE_OK) {
                        CPP_REPORT(result, "Could not apply TopicQos.");
                    }
                }
            }
            this->unlock ();
        }

        u_topicQosFree (uTopicQos);
    }

    CPP_REPORT_FLUSH(this, result != DDS::RETCODE_OK);

    return result;
}

DDS::TopicListener_ptr
DDS::OpenSplice::Topic::get_listener (
) THROW_ORB_EXCEPTIONS
{
    DDS::TopicListener_ptr listener;

    listener = dynamic_cast<DDS::TopicListener_ptr>(
        DDS::OpenSplice::Entity::nlReq_get_listener());

    return listener;
}

DDS::ReturnCode_t
DDS::OpenSplice::Topic::set_listener (
    DDS::TopicListener_ptr a_listener,
    DDS::StatusMask mask
) THROW_ORB_EXCEPTIONS
{
    DDS::ReturnCode_t result;

    CPP_REPORT_STACK();

    result = this->write_lock ();
    if (result == DDS::RETCODE_OK) {
        result = DDS::OpenSplice::Entity::wlReq_set_listener(a_listener, mask | participantListenerInterest);
        if (result == DDS::RETCODE_OK) {
            if (a_listener != NULL) {
                topicListenerInterest = mask;
            } else {
                topicListenerInterest = 0;
            }
        }
        this->unlock ();
    }

    CPP_REPORT_FLUSH(this, result != DDS::RETCODE_OK);

    return result;
}

DDS::ReturnCode_t
DDS::OpenSplice::Topic::set_participant_listener_mask (
    ::DDS::StatusMask mask
) THROW_ORB_EXCEPTIONS
{
    DDS::ReturnCode_t result;

    CPP_REPORT_STACK();

    result = this->write_lock ();
    if (result == DDS::RETCODE_OK) {
        result = DDS::OpenSplice::Entity::wlReq_set_listener_mask(mask | topicListenerInterest);
        if (result == DDS::RETCODE_OK) {
            participantListenerInterest = mask;
        }
        this->unlock ();
    }

    CPP_REPORT_FLUSH(this, result != DDS::RETCODE_OK);

    return result;
}

DDS::ReturnCode_t
DDS::OpenSplice::Topic::dispose_all_data (
) THROW_ORB_EXCEPTIONS
{
    DDS::ReturnCode_t result;
    u_topic uTopic;
    u_result uResult;

    CPP_REPORT_STACK();

    result = this->write_lock ();
    if (result == DDS::RETCODE_OK) {
        uTopic = u_topic(DDS::OpenSplice::Entity::rlReq_get_user_entity());
        uResult = u_topicDisposeAllData(uTopic);
        result = uResultToReturnCode (uResult);
        if (result != DDS::RETCODE_OK) {
            CPP_REPORT(result, "Could not dispose all data for Topic.");
        }
        this->unlock ();
    }

    CPP_REPORT_FLUSH(this, result != DDS::RETCODE_OK);

    return result;
}

void
DDS::OpenSplice::Topic::nlReq_notify_listener(
    DDS::OpenSplice::Entity *sourceEntity,
    DDS::ULong               triggerMask,
    void                    *eventData)
{
    DDS::TopicListener_ptr listener;

    /* Using _narrow to cast Listener, this increases the refcount to ensure
     * the ListenerObject is not deleted while notifying. */
    listener = DDS::TopicListener::_narrow(this->listener);
    if (triggerMask & V_EVENT_INCONSISTENT_TOPIC) {
        if (listener && (topicListenerInterest & DDS::INCONSISTENT_TOPIC_STATUS)) {
            DDS::InconsistentTopicStatus status;
            DDS::OpenSplice::Utils::copyStatusOut(v_topicStatus(eventData)->inconsistentTopic, status);
            listener->on_inconsistent_topic(dynamic_cast<DDS::Topic_ptr>(sourceEntity), status);
        } else if (participantListenerInterest & DDS::INCONSISTENT_TOPIC_STATUS) {
            participant->nlReq_notify_listener(sourceEntity, triggerMask, eventData);
        }
    }

    if (triggerMask & V_EVENT_ALL_DATA_DISPOSED) {
        if (listener && (topicListenerInterest & DDS::ALL_DATA_DISPOSED_TOPIC_STATUS)) {
            DDS::ExtTopicListener_ptr extListener;
            extListener = dynamic_cast<DDS::ExtTopicListener_ptr>(listener);
            if (extListener) {
                extListener->on_all_data_disposed(dynamic_cast<DDS::Topic_ptr>(sourceEntity));
            }
        } else if (participantListenerInterest & DDS::ALL_DATA_DISPOSED_TOPIC_STATUS) {
            participant->nlReq_notify_listener(sourceEntity, triggerMask, eventData);
        }
    }
    DDS::release(listener);
}
